File: itmo/contest_6/F6.py
import sys
from math import floor


def insertion_sort(arr):
    if len(arr) == 1:
        return

    for i in range(1, len(arr)):
        key = arr[i]
        j = i - 1
        while j >= 0 and arr[j] > key:
            arr[j + 1] = arr[j]
            j -= 1
        arr[j + 1] = key


def main():
    t, n = map(int, sys.stdin.readline().split())
    a, b = map(int, sys.stdin.readline().split())

    cur = 0
    buckets = [[] for _ in range(n)]
    const = n / 4294967296

    for _ in range(t):
        idx, summ = 1, 0

        for _ in range(n):
            cur = (cur * a + b) % 4294967296
            x = cur >> 8
            cur = (cur * a + b) % 4294967296
            y = cur >> 8
            z = ((x << 8) ^ y) % 4294967296

            buckets[floor(const * z)].append(z)

        for bucket in buckets:
            insertion_sort(bucket)
            for x in bucket:
                summ += x * idx
                idx += 1
            del bucket[:]
        print(summ)


if __name__ == "__main__":
    main()

File: itmo/contest_6/F6.cpp
#include <stdint.h>
#include <iostream>
#include <vector>
#include <cmath>
#include <algorithm>

using namespace std;

uint32_t a, b;
uint32_t cur = 0;

uint32_t nextRand24() {
    cur = cur * a + b;
    return cur >> 8;
}

uint32_t nextRand32() {
    uint32_t x = nextRand24(), y = nextRand24();
    return (x << 8) ^ y;
}

void insertion_sort(vector<uint32_t>& arr) {
    for (int i = 1; i < arr.size(); ++i) {
        uint32_t key = arr[i];
        int j = i - 1;
        while (j >= 0 && arr[j] > key) {
            arr[j + 1] = arr[j];
            j--;
        }
        arr[j + 1] = key;
    }
}

int main() {
    int t, n;
    cin >> t >> n;
    cin >> a >> b;

    vector<vector<uint32_t>> buckets(n);
    double const_value = static_cast<double>(n) / 4294967296.0;

    for (int _ = 0; _ < t; ++_) {
        int idx = 1;
        long long summ = 0;

        for (auto& bucket : buckets) {
            bucket.clear();
        }

        for (int i = 0; i < n; ++i) {
            uint32_t z = nextRand32();
            int bucket_idx = static_cast<int>(floor(const_value * z));
            buckets[bucket_idx].push_back(z);
        }

        for (auto& bucket : buckets) {
            insertion_sort(bucket);

            for (uint32_t x : bucket) {
                summ += static_cast<long long>(x) * idx;
                idx++;
            }
        }

        cout << summ << endl;
    }

    return 0;
}
